### PR TITLE
changed to use 'advance_read()' instead of 'end_read()' and 'begin_read()' when possible.

### LINKED ISSUE #2470: We can reduce the chance of `BadVersionException` in some cases
https://github.com/realm/realm-java/blob/5444c96b8bec2a6537f30401cca3b452aab1a42a/realm/realm-jni/src/io_realm_internal_TableQuery.cpp#L1055

https://github.com/realm/realm-java/blob/5444c96b8bec2a6537f30401cca3b452aab1a42a/realm/realm-jni/src/io_realm_internal_TableQuery.cpp#L1166

```
    } else if (sg->get_version_of_current_transaction() != handoverQuery->version) {
        sg->end_read();
        sg->begin_read(handoverQuery->version);
    }
```

The code above can be:

```
    } else if (sg->get_version_of_current_transaction() < handoverQuery->version) {
        sg->advance_read(handoverQuery->version);
    } else if (handoverQuery->version < sg->get_version_of_current_transaction()) {
        sg->end_read();
        sg->begin_read(handoverQuery->version);
    }
```

By avoiding to call `end_read()`, we can reduce the chance of `BadVersionException`.

### PR DESCRIPTION
**Pull request by zaki50**
This fixes #2470

### CODE REVIEW

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 1]
```diff
@@ -1051,7 +1051,11 @@ std::unique_ptr<Query> handoverQueryToWorker(jlong bgSharedGroupPtr, jlong query
     if (sg->get_transact_stage() != SharedGroup::transact_Reading) {
         // if the SharedGroup is not in Read Transaction, we position it at the same version as the handover
         sg->begin_read(handoverQuery->version);
-    } else if (sg->get_version_of_current_transaction() != handoverQuery->version) {
+    } else if (sg->get_version_of_current_transaction() < handoverQuery->version) {
+        // going forward
+        LangBindHelper::advance_read(*sg, handoverQuery->version);
+    } else if (handoverQuery->version < sg->get_version_of_current_transaction()) {
+        // going backward
```
Comment (zaki50) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: I think we should not go backward and it's enough to leave the version as it is.

What do you think? @nhachicha @realm/java

Comment (cmelchior) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: I'll need to some time to think through the flow of this, especially regarding where the SharedGroup is actually created in the flow.

**Comment by beeender** [comment]
Any test result to show how much it reduce the possibility of generating bad version exception?

**Comment by zaki50** [comment]
@beeender I'll try to make the testcase.

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp (hunk 1 above)
Comment (nhachicha) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: we do need to go backward, this is most of the time what happens, when you receive a `REALM_CHANGE` an you try to update your query, obviously the UI thread delayed the SharedGroup advance, until the async queries updates. so the handed over query will most likely be behind the latest worker thread shared group

**Comment by cmelchior** [comment]
@zaki50 Given the discussion above. This can just be closed, right?

**Comment by zaki50** [comment]
@cmelchior yes.

**Pull request closed without merging**